website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md

```diff
@@ -0,0 +1,160 @@
+For instance, it could be a BSON document with particular tracing-related fields.
+This would provide a more reliable method for passing context to the database.
+
+## Example application
+
+Since such a standard does not yet exist, let's explore [an example application](https://gist.github.com/rumyantseva/3c6ef7c7dfc3fbdea8f94a31f4a17885)
+that interacts with FerretDB and passes the tracing context using the `comment` field.
+
+For simplicity, most error handling is omitted in the example below.
+
+```go
+collection := client.Database("testdb").Collection("customers")
+
+user := bson.D{
+    {Key: "name", Value: "John Doe"},
+    {Key: "email", Value: "john.doe@example.com"},
+}
+
+insertCtx, insertSpan := tracer.Start(ctx, "InsertCustomer")
+_, _ = collection.InsertOne(insertCtx, user)
+insertSpan.End()
+
+findCtx, findSpan := tracer.Start(ctx, "FindCustomer")
+filter := bson.D{{Key: "name", Value: "John Doe"}}
+
+var result bson.D
+_ = collection.FindOne(findCtx, filter).Decode(&result)
+
+findSpan.End()
+```
+
+In this setup, we create two spans: one for inserting a document and another for finding it.
+
+If we run this application, we will see that the spans for the `InsertCustomer` and `FindCustomer` operations are created:
+
+![Trace without propagation](/img/blog/ferretdb-otel/without-propagation.png)
+
+Let's add more details.
+If we pass the tracing context through the `comment` field, we will see that the spans are linked.
+Let's modify the `FindCustomer` part:
+
+```go
+findCtx, findSpan := tracer.Start(ctx, "FindCustomer")
+filter := bson.D{{Key: "name", Value: "John Doe"}}
+
+traceID := findSpan.SpanContext().TraceID().String()
+spanID := findSpan.SpanContext().SpanID().String()
+
+traceContext := struct {
+    TraceID string `json:"traceID"`
+    SpanID  string `json:"spanID"`
+}{
+    TraceID: traceID,
+    SpanID:  spanID,
+}
+
+comment, _ := json.Marshal(map[string]interface{}{
+    "ferretDB": traceContext,
+})
+
+var result bson.D
+_ = collection.FindOne(findCtx, filter, options.FindOne().SetComment(string(comment))).Decode(&result)
+
+findSpan.End()
+```
+
+Now, if we run the application, we will see that the spans are linked and the `FindCustomer` span is a child span created on the FerretDB side:
+
+![Trace with propagation](/img/blog/ferretdb-otel/with-propagation.png)
+
+This approach gives more insights into the FerretDB's behavior and helps to understand the exact query lifecycle,
+making it easier to diagnose and understand performance issues or unexpected behavior.
+
+While this solution isn't perfect due to the limitations discussed, it is a step towards better context propagation in document databases.
+
+## Conclusion
+
+We believe that passing context to document databases is an important part of making them more observable.
+We hope that the community will come up with a standard way to pass context to databases, and we are looking forward to
+contributing to this effort.
+
```

Comment:
https://github.com/FerretDB/FerretDB/pull/4512#pullrequestreview-2235093373 might be somewhere there